sodetlib/operations/uxm_setup.py

```diff
@@ -363,6 +363,162 @@ def setup_tune(S, cfg, bands, show_plots=False, update_cfg=True):
+
+    # return the central frequencies in selected gaps
+    gap_ind = np.argsort(gaps)[:-(ntone + 1):-1]  # ntone largest gaps
+    return 0.5 * (res_freq[gap_ind + 1] + res_freq[gap_ind])
+
+
+@sdl.set_action()
+def setup_fixed_tones(
+    S, cfg, fixed_tones_per_band=8, min_gap=10, tone_power=10, update_cfg=True
+    ):
+    """
+    Identify gaps between resonators and attempt to place a number of fixed tones there.
+    These will be saved for every band in the device config. A tune should be available
+    before running this action.
+
+    Args
+    ----
+    S : SmurfControl
+        Pysmurf instance
+    cfg : DetConfig
+        DetConfig instance
+    fixed_tones_per_band : int
+        The number of tones to place in each band. Defaults to 8.
+    min_gap : float
+        The minimum gap size in order to place a fixed tone, in MHz. Defaults to 10.
+    tone_power : int
+        The tone power to use for the fixed tones. Defaults to 10.
+    update_cfg : bool
+        Whether to save the fixed tone configuration to the device config. Defaults to True.
+
+    Returns
+    -------
+    fixed_tones : dict
+        Fixed tones configuration, as saved to device config.
+    """
+
+    bands = cfg.dev.exp['active_bands']
+
+    # ensure we have a tune available
+    try:
+        _ = S.freq_resp[bands[0]]["resonances"]
+    except KeyError as e:
+        raise ValueError("No tune is available to set fixed tones from.") from e
+
+    # identify available frequencies
+    ft_freq = []
+    for b in bands:
+        ft_freq += list(_find_fixed_tone_freq(S, b, fixed_tones_per_band, min_gap))
+
+        # turn off any existing fixed tones
+        try:
+            channels = cfg.dev.bands[b]["fixed_tones"]["channels"]
+            amp = S.get_amplitude_scale_array(b)
+            amp[channels] = 0.0
+            S.set_amplitude_scale_array(b, amp)
+            S.log(f"Turned off current fixed tones on band {b}.") 
+        except KeyError:
+            pass
+
+    # set the fixed tones
+    fixed_tones = {
+        b: {"freq_offsets": [], "channels": [], "tone_power": tone_power}
+        for b in bands
+    }
+    for fr in ft_freq:
+        # this function sets the tone power and disables feedback
+        try:
+            band, channel = S.set_fixed_tone(fr, tone_power=tone_power)
+        except AssertionError as e:
+            S.log(f"Failed to assign a fixed tone on frequency {fr} MHz.\n{e}.")
+            continue
+        # read back the assigned frequency offset
+        foff = S.get_center_frequency_mhz_channel(band, channel)
+        fixed_tones[band]["freq_offsets"].append(foff)
+        fixed_tones[band]["channels"].append(channel)
+
+    # update config
+    if update_cfg:
+        for band, data in fixed_tones.items():
+            cfg.dev.update_band(band, {"fixed_tones": data})
+
+    S.log("Done setting fixed tones.")
+
+    return fixed_tones
+
+
+@sdl.set_action()
+def turn_on_fixed_tones(S, cfg, band):
```

Comment:
Can this also default to running on a group of bands (defaulting to active_bands) instead of just one? That is how we will always be running in operation so its easier if we don't need to loop over bands every time.